src/fondant/pipeline/lightweight_component.py

```diff
@@ -110,6 +126,41 @@ def validate_abstract_methods_are_implemented(cls):
                     msg,
                 )
 
+        def modify_consumes_spec(apply_consumes, consumes_spec):
```

Comment:
Isn't this functionality already available in the `OperationSpec` class? I would try to keep the implementation in line with the custom components:

- Build the consumes specification based on the decorator `consumes`
- Leverage both the consumes specification and consumes apply argument in the `OperationSpec`

Reply:
so the flow currently goes like [this](https://github.com/ml6team/fondant/pull/789/files#diff-65a733e81e8bc30179d9f957e51a2c6e9e45bc2a4101d54d6d5072f98433a69aR222-R246:~:text=image%20%3D%20ref,**kwargs%2C) 
-> get `consumes_spec` (remap names if necessary and filter) -> pass it to the `consumes` section of the component spec (actual consumed fields) -> build the `OperationOp` 


The reason why we're modifying things in the decorator is that the component spec should already have the mapped name, in resuable components this is usually the case but here it's a bit different. 

Let's take an example where we have a dataset of field `x`  and the next component accepts a field `a`

The component spec for the reusable component would already have the field `a` in it's schema. For lightweight components, we start with the dataset schema (`x`) that needs to be remapped to `a` before being passed to the component spec and then the componentOp.

To replicate the issue you can comment out the function `modify_consumes_spec` and run [this test](https://github.com/ml6team/fondant/pull/789/files#diff-ea78d89d3d049a673b92c8463cce4bf08fa68fa52e677b7fc1cf95f0abdfe8a5R145:~:text=(pipeline)-,def%20test_valid_consumes_mapping(tmp_path_factory%2C%20load_pipeline)%3A,assert,-%22z%22%20not). By commenting out the mapping and filter function 

```python
@classmethod
def get_consumes_spec(
    cls,
    dataset_fields: t.Mapping[str, Field],
    apply_consumes: t.Optional[t.Dict[str, t.Union[str, pa.DataType]]],
):
    python_component_consumes = cls.consumes()

    # Get consumes spec from the dataset
    consumes_spec = {k: v.type.to_dict() for k, v in dataset_fields.items()}

    # Modify naming based on the consumes argument in the 'apply' method
    #\consumes_spec = cls.modify_consumes_spec(apply_consumes, consumes_spec)

    # Filter for values that are not in the user defined consumes list
    # consumes_spec = cls.filter_consumes_spec(
    #     python_component_consumes,
    #     consumes_spec,
    # )

    return consumes_spec
```

```
E  fondant.core.exceptions.InvalidPipelineDefinition: Received a string value for key `a` in the `consumes` argument passed to the operation, but `a` is not defined in the `consumes` section of the component spec.
```

Reply:
The flow I suggest is the following:

```python
# First build the equivalent of the component spec
consumes_spec = create_consumes_spec(dataset_fields, python_component_consumes)
component_spec = ComponentSpec(consumes_spec, ...)

# Then apply the same logic as for containerized components
operation_spec = OperationSpec(component_spec, apply_consumes)
```

Reply:
I think the issue will still persist: 

* The component spec should always contain the names of the columns that will be consumed by the component 
* This was not an issue before since we have it explicitly written in a yaml 
* Now we are starting from the dataset schema/fields which may have different names than the ones expected by the component, thus we need to dynamically infer the component spec.
* The `dataset_fields` and `python_component_consumes` which just contains a list of columns to consume alone are not enough to get the ground truth component spec, for that we need to remap it using the `apply` consumes before passing it to the `OperationSpec` in order to get the to the actual component spec expected by the component. 

For example, starting from a dataset_fields of x,y and assuming `python_component_consumes` is [`x`], we cannot infer the component spec for the next component which may expect `z` -> leads to errors.

I think that's the main reason why we need to include some custom logic beforehand